File: 20-valid-parenthesis/smart.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
  if (s.length % 2 === 1) return false
  
  const stack = []
  const bracesMap = {
    '{': '}',
    '(': ')',
    '[': ']',
  }
  let expectedClosingBrace = ''
  let isValid = true
  
  s.split('').forEach(brace => {
    if (isValid) {
      currentClosingBrace = bracesMap[brace]
      
      if (currentClosingBrace) {
        stack.push(brace)
        expectedClosingBrace = currentClosingBrace
        return
      }

      if (brace === expectedClosingBrace) {
        stack.pop()
        expectedClosingBrace = bracesMap[stack[stack.length - 1]]
      } else {
        isValid = false
      }
    }
  })
  
  if (!isValid || stack.length) return false
  else return true
}

File: 20-valid-parenthesis/smart.go
package validparenthesis

func isValid(s string) bool {
	strLen := len(s)

	// if s has odd length, then it definitely isnt valid
	if strLen%2 == 1 {
		return false
	}

	stack := make([]string, 0, strLen)
	bracesMap := map[string]string{
		"{": "}",
		"(": ")",
		"[": "]",
	}
	expectedClosingBrace := ""

	for _, brace := range s {
		if matchingClosingBrace, ok := bracesMap[string(brace)]; ok {
			stack = append(stack, string(brace))
			expectedClosingBrace = matchingClosingBrace
			continue
		}
		if string(brace) == expectedClosingBrace {
			stack = stack[:len(stack)-1]
			if len(stack) > 0 {
				expectedClosingBrace = bracesMap[stack[len(stack)-1]]
			} else {
				expectedClosingBrace = ""
			}
		} else {
			return false
		}
	}

	if len(stack) == 0 {
		return true
	} else {
		return false
	}
}
